File: SharedMemory/shared_memory_client.cpp
#include <cstring>
#include <fcntl.h>
#include <iostream>
#include <sys/mman.h>
#include <unistd.h>

constexpr const char *SHM_NAME = "/my_shared_memory";
constexpr size_t SHM_SIZE = 1024 * 1024;

int main(int argc, char *argv[]) {
  // Open SHM
  int shm_fd = shm_open(SHM_NAME, O_RDONLY, 0666);
  void *shm_ptr = mmap(nullptr, SHM_SIZE, PROT_READ, MAP_SHARED, shm_fd, 0);

  size_t offset = 0;
  size_t key_len;
  memcpy(&key_len, static_cast<char *>(shm_ptr) + offset, sizeof(key_len));
  offset += sizeof(key_len);
  std::string key(static_cast<char *>(shm_ptr) + offset, key_len);
  offset += key_len;

  size_t value_len;
  memcpy(&value_len, static_cast<char *>(shm_ptr) + offset, sizeof(value_len));
  offset += sizeof(value_len);
  std::string value(static_cast<char *>(shm_ptr) + offset, value_len);
  offset += value_len;

  std::cout << "Retrieved '" << key << "' with value '" << value
            << "' from shared memory." << std::endl;

  return 0;
}

File: SharedMemory/shared_memory_client.py
import mmap
import struct
import os

SHM_NAME = '/my_shared_memory'
SHM_SIZE = 1024 * 1024

# In Linux, POSIX shared memory objects are in /dev/shm directory
shm_path = f"/dev/shm{SHM_NAME}"

# Check if the shared memory exists
if not os.path.exists(shm_path):
    raise FileNotFoundError(f"Shared memory {SHM_NAME} not found. Is the server running?")

with open(shm_path, 'r+b') as f:
    with mmap.mmap(f.fileno(), SHM_SIZE, access=mmap.ACCESS_READ) as shm:
        offset = 0
        # Use 'Q' for 8-byte size_t on 64-bit systems (or 'I' for 4-byte on 32-bit systems)
        key_length = struct.unpack('Q', shm[offset:offset + 8])[0]
        offset += 8
        key = shm[offset:offset + key_length].decode('utf-8')
        offset += key_length
        
        value_length = struct.unpack('Q', shm[offset:offset + 8])[0]
        offset += 8
        value = shm[offset:offset + value_length].decode('utf-8')
        print(f"Retrieved '{key}' with value '{value}' from shared memory.")
